common/src/ethereum/walletconnect.rs

```diff
@@ -0,0 +1,104 @@
+use std::error::Error;
+use walletconnect::client::{CallError, SessionError};
+use walletconnect::{qr, Client, Metadata, Transaction};
+
+use ethers::types::transaction::eip2718::TypedTransaction;
+use ethers::types::{Address, NameOrAddress, Signature, U256};
+
+#[derive(Debug)]
+pub struct WalletConnect {
+    client: Client,
+    chain_id: u64,
+    address: Address,
+}
+
+impl WalletConnect {
+    pub fn new() -> Result<Self, Box<dyn Error>> {
+        let client = Client::new(
+            "rad-ens",
+            Metadata {
+                description: "Change ENS records for your Radicle profile".into(),
+                url: "https://github.com/radicle-dev".parse()?,
+                icons: vec!["https://avatars0.githubusercontent.com/u/48290027".parse()?],
+                name: "Radicle CLI".into(),
+            },
+        )?;
+
+        Ok(WalletConnect {
+            client,
+            chain_id: 0,
+            address: Address::zero(),
+        })
+    }
+
+    pub async fn show_wait_connection(mut self) -> Result<Self, SessionError> {
+        let (accounts, chain_id) = self
+            .client
+            .ensure_session(|uri| {
+                println!("{}", uri.as_url());
+                qr::print(uri);
+            })
+            .await?;
+
+        self.chain_id = chain_id;
+        self.address.assign_from_slice(accounts[0].as_bytes());
+
+        Ok(self)
+    }
+
+    pub fn chain_id(&self) -> u64 {
+        self.chain_id
+    }
+
+    pub fn address(&self) -> Address {
+        self.address
+    }
+
+    fn address_string(&self) -> String {
+        format!("{}", self.address)
+    }
+
+    pub async fn sign_message<S: Send + Sync + AsRef<[u8]>>(
+        &self,
+        msg: S,
+    ) -> Result<Signature, CallError> {
+        let msg = unsafe { std::str::from_utf8_unchecked(msg.as_ref()) };
```

Comment:
Can we take a `&str` or `String` as `msg` instead of having to convert it here?

Reply:
Don't think so, tried at first but couldn't get it to compile.

Reply:
But `msg: &str` here, so can't you just take `&str` as the function param?

Reply:
You mean like this?

```
    pub async fn sign_message(&self, msg: &str) -> Result<Signature, CallError> {
        self.client
            .personal_sign(&[msg, &self.address_string()])
            .await
    }
```

Issue is downstream:
```
error[E0308]: mismatched types
   --> common/src/ethereum.rs:204:54
    |
197 |     async fn sign_message<S: Send + Sync + AsRef<[u8]>>(
    |                           - this type parameter
...
204 |             Self::WalletConnect(s) => s.sign_message(message).await.map_err(WalletError::from),
    |                                                      ^^^^^^^ expected `&str`, found type parameter `S`
    |
    = note:   expected reference `&str`
            found type parameter `S`
note: return type inferred to be `&str` here
   --> common/src/ethereum.rs:200:41
    |
200 |       ) -> Result<Signature, Self::Error> {
    |  _________________________________________^
201 | |         match self {
202 | |             Self::Ledger(s) => s.sign_message(message).await.map_err(WalletError::from),
203 | |             Self::Local(s) => s.sign_message(message).await.map_err(WalletError::from),
204 | |             Self::WalletConnect(s) => s.sign_message(message).await.map_err(WalletError::from),
205 | |         }
206 | |     }
    | |_____^
```

Reply:
Oh I see. Ok, I'll look into this later.